### PR TITLE
Implement highlight on hover using a Color Picking technique

### PR DESCRIPTION
**Pull request by martinRenou**
For now this is only a visual effect, the state is not synced between clients. We can discuss in an issue whether this could be useful or not.

https://user-images.githubusercontent.com/21197331/194020360-5515801c-fee8-4d9d-99ef-f8d40dcdc568.mp4

### CODE REVIEW

**Comment by trungleduc** [comment]
Thanks @martinRenou, It looks great! 
For sharing state with other users, I don't think it is useful but is it possible to highlight an object when we select it from the control panel?

**Comment by martinRenou** [comment]
So actually, maybe instead of an "highlight on hover" behavior, it would make sense to implement a click selection, and when you click on the piece it selects it in the control panel?

**Comment by trungleduc** [comment]
Freecad has both hover and selection events, but I think the selection is more useful.

fc

> That way we can both select from the control panel and select from the 3D view.

That's great!

**Comment by martinRenou** [comment]
Let's do just like FreeCad! The color-picking being super cheap, I think we can afford it.

**Comment by martinRenou** [comment]
I implemented both an hover logic and a click logic for selection:

https://user-images.githubusercontent.com/21197331/194053313-84059174-b6af-4217-af0d-5cc29fb7ba17.mp4

We can bind the control panel UI to the actual 3D selection in a separate PR

File: src/mainview.tsx [hunk 1]
```diff
@@ -250,10 +270,79 @@ export class MainView extends React.Component<IProps, IStates> {
...
+    this._renderer.readRenderTargetPixels(
+      this._pickingTexture,
+      0,
+      0,
+      1,
+      1,
+      pixelBuffer
+    );
+
+    const id = (pixelBuffer[0] << 16) | (pixelBuffer[1] << 8) | pixelBuffer[2];
```
Comment (trungleduc) [inline src/mainview.tsx:301]: I'm just curious, what does this line mean?

File: src/mainview.tsx [hunk 2]
```diff
@@ -301,13 +390,26 @@ export class MainView extends React.Component<IProps, IStates> {
...
-    if (old) {
-      this._scene.remove(old);
+    if (this._meshGroup !== null) {
+      this._scene.remove(this._meshGroup);
     }
-    const mainObject = new THREE.Group();
-    mainObject.name = 'shape';
+
+    const pickingColor = new THREE.Color();
+    let i = 1;
```
Comment (trungleduc) [inline src/mainview.tsx:410]:
```suggestion
```

File: src/mainview.tsx [hunk 3]
```diff
@@ -301,13 +390,26 @@ export class MainView extends React.Component<IProps, IStates> {
...
+
+    this._meshGroup = new THREE.Group();
     Object.entries(payload).forEach(([objName, data]) => {
```
Comment (trungleduc) [inline src/mainview.tsx:413]:
```suggestion
    Object.entries(payload).forEach(([objName, data], idx) => {
```

File: src/mainview.tsx [hunk 4]
```diff
@@ -366,6 +470,12 @@ export class MainView extends React.Component<IProps, IStates> {
       model.castShadow = true;
       model.name = objName;
 
+      // Add the mesh to the color picking scene
+      const pickingMaterial = new THREE.MeshBasicMaterial({
+        color: pickingColor.setHex(i)
```
Comment (trungleduc) [inline src/mainview.tsx:475]:
```suggestion
        color: pickingColor.setHex(idx + 1)
```

File: src/mainview.tsx [hunk 5]
```diff
@@ -380,10 +490,12 @@ export class MainView extends React.Component<IProps, IStates> {
 
       model.add(mesh);
 
-      mainObject.add(model);
+      this._meshGroup!.add(model);
+
+      i++;
```
Comment (trungleduc) [inline src/mainview.tsx:495]:
```suggestion
```

File: src/mainview.tsx (hunk 2 above)
Comment (trungleduc) [reply src/mainview.tsx:410]: I think we can replace this counter with the index of the `forEach` loop.

File: src/mainview.tsx (hunk 1 above)
Comment (trungleduc) [reply src/mainview.tsx:301]: Thanks!

**Review by trungleduc — APPROVED** [review]

**Comment by trungleduc** [comment]
Thanks!

**Pull request merged**